File: java/helidon/src/main/java/org/openapitools/server/db/PetRepository.java
package org.openapitools.server.db;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import jakarta.ws.rs.WebApplicationException;
import jakarta.ws.rs.core.Response;
import org.openapitools.server.model.Category;
import org.openapitools.server.model.Pet;
import org.openapitools.server.model.Tag;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Types;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicLong;

@ApplicationScoped
public class PetRepository {

    @Inject
    DataSourceProvider dsProvider;

    private static final AtomicLong idGen = new AtomicLong(System.currentTimeMillis());
    private final ObjectMapper mapper = new ObjectMapper();

    public Pet add(Pet pet) {
        try (Connection conn = dsProvider.get().getConnection()) {
            long id = pet.getId() != null ? pet.getId() : idGen.getAndIncrement();
            String sql =
                "INSERT INTO pet (\"id\", \"name\", category, photo_urls, tags, status) " +
                "VALUES (?, ?, ?, cast(? as json), cast(? as json), cast(? as pet_status))";
            try (PreparedStatement ps = conn.prepareStatement(sql)) {
                ps.setLong(1, id);
                ps.setString(2, pet.getName());
                ps.setString(3, pet.getCategory() != null ? mapper.writeValueAsString(pet.getCategory()) : null);
                ps.setString(4, mapper.writeValueAsString(pet.getPhotoUrls()));
                ps.setString(5, pet.getTags() != null ? mapper.writeValueAsString(pet.getTags()) : null);
                ps.setString(6, pet.getStatus() != null ? pet.getStatus().value() : null);
                ps.executeUpdate();
            }
            pet.setId(id);
            return pet;
        } catch (WebApplicationException e) {
            throw e;
        } catch (Exception e) {
            throw new WebApplicationException("Failed to add pet: " + e.getMessage(),
                Response.Status.INTERNAL_SERVER_ERROR);
        }
    }

    public Pet update(Pet pet) {
        if (pet.getId() == null) {
            throw new WebApplicationException("Pet ID is required for update", Response.Status.BAD_REQUEST);
        }
        try (Connection conn = dsProvider.get().getConnection()) {
            String sql =
                "UPDATE pet SET \"name\" = ?, category = ?, " +
                "photo_urls = cast(? as json), tags = cast(? as json), status = cast(? as pet_status) " +
                "WHERE \"id\" = ?";
            try (PreparedStatement ps = conn.prepareStatement(sql)) {
                ps.setString(1, pet.getName());
                ps.setString(2, pet.getCategory() != null ? mapper.writeValueAsString(pet.getCategory()) : null);
                ps.setString(3, mapper.writeValueAsString(pet.getPhotoUrls()));
                ps.setString(4, pet.getTags() != null ? mapper.writeValueAsString(pet.getTags()) : null);
                ps.setString(5, pet.getStatus() != null ? pet.getStatus().value() : null);
                ps.setLong(6, pet.getId());
                if (ps.executeUpdate() == 0) {
                    throw new WebApplicationException("Pet not found", Response.Status.NOT_FOUND);
                }
            }
            return pet;
        } catch (WebApplicationException e) {
            throw e;
        } catch (Exception e) {
            throw new WebApplicationException("Failed to update pet: " + e.getMessage(),
                Response.Status.INTERNAL_SERVER_ERROR);
        }
    }

    public Pet findById(long petId) {
        try (Connection conn = dsProvider.get().getConnection()) {
            String sql =
                "SELECT \"id\", \"name\", category, photo_urls, tags, status::text " +
                "FROM pet WHERE \"id\" = ?";
            try (PreparedStatement ps = conn.prepareStatement(sql)) {
                ps.setLong(1, petId);
                try (ResultSet rs = ps.executeQuery()) {
                    if (rs.next()) {
                        return mapRow(rs);
                    }
                }
            }
            throw new WebApplicationException("Pet not found", Response.Status.NOT_FOUND);
        } catch (WebApplicationException e) {
            throw e;
        } catch (Exception e) {
            throw new WebApplicationException("Failed to get pet: " + e.getMessage(),
                Response.Status.INTERNAL_SERVER_ERROR);
        }
    }

    public void delete(long petId) {
        try (Connection conn = dsProvider.get().getConnection()) {
            try (PreparedStatement ps = conn.prepareStatement("DELETE FROM pet WHERE \"id\" = ?")) {
                ps.setLong(1, petId);
                ps.executeUpdate();
            }
        } catch (Exception e) {
            throw new WebApplicationException("Failed to delete pet: " + e.getMessage(),
                Response.Status.INTERNAL_SERVER_ERROR);
        }
    }

    public List<Pet> findByStatus(String status) {
        try (Connection conn = dsProvider.get().getConnection()) {
            String sql =
                "SELECT \"id\", \"name\", category, photo_urls, tags, status::text " +
                "FROM pet WHERE status = cast(? as pet_status)";
            try (PreparedStatement ps = conn.prepareStatement(sql)) {
                ps.setString(1, status);
                try (ResultSet rs = ps.executeQuery()) {
                    List<Pet> pets = new ArrayList<>();
                    while (rs.next()) {
                        pets.add(mapRow(rs));
                    }
                    return pets;
                }
            }
        } catch (Exception e) {
            throw new WebApplicationException("Failed to find pets by status: " + e.getMessage(),
                Response.Status.INTERNAL_SERVER_ERROR);
        }
    }

    public List<Pet> findByTags(List<String> tags) {
        if (tags == null || tags.isEmpty()) {
            return new ArrayList<>();
        }
        try (Connection conn = dsProvider.get().getConnection()) {
            StringBuilder sql = new StringBuilder(
                "SELECT \"id\", \"name\", category, photo_urls, tags, status::text FROM pet WHERE false");
            for (int i = 0; i < tags.size(); i++) {
                sql.append(" OR tags::jsonb @> cast(? as jsonb)");
            }
            try (PreparedStatement ps = conn.prepareStatement(sql.toString())) {
                for (int i = 0; i < tags.size(); i++) {
                    ObjectNode node = mapper.createObjectNode().put("name", tags.get(i));
                    ArrayNode arr = mapper.createArrayNode().add(node);
                    ps.setString(i + 1, mapper.writeValueAsString(arr));
                }
                try (ResultSet rs = ps.executeQuery()) {
                    List<Pet> pets = new ArrayList<>();
                    while (rs.next()) {
                        pets.add(mapRow(rs));
                    }
                    return pets;
                }
            }
        } catch (Exception e) {
            throw new WebApplicationException("Failed to find pets by tags: " + e.getMessage(),
                Response.Status.INTERNAL_SERVER_ERROR);
        }
    }

    public void updateWithForm(long petId, String name, String status) {
        if (name == null && status == null) {
            return;
        }
        try (Connection conn = dsProvider.get().getConnection()) {
            StringBuilder sql = new StringBuilder("UPDATE pet SET ");
            List<Object> params = new ArrayList<>();
            if (name != null) {
                sql.append("\"name\" = ?");
                params.add(name);
            }
            if (status != null) {
                if (!params.isEmpty()) sql.append(", ");
                sql.append("status = cast(? as pet_status)");
                params.add(status);
            }
            sql.append(" WHERE \"id\" = ?");
            params.add(petId);
            try (PreparedStatement ps = conn.prepareStatement(sql.toString())) {
                for (int i = 0; i < params.size(); i++) {
                    ps.setObject(i + 1, params.get(i));
                }
                ps.executeUpdate();
            }
        } catch (Exception e) {
            throw new WebApplicationException("Failed to update pet with form: " + e.getMessage(),
                Response.Status.INTERNAL_SERVER_ERROR);
        }
    }

    public Map<String, Integer> getInventory() {
        try (Connection conn = dsProvider.get().getConnection()) {
            String sql = "SELECT status::text, cast(COUNT(*) as int) FROM pet GROUP BY status";
            try (PreparedStatement ps = conn.prepareStatement(sql);
                 ResultSet rs = ps.executeQuery()) {
                Map<String, Integer> inventory = new HashMap<>();
                while (rs.next()) {
                    String s = rs.getString(1);
                    if (s != null) {
                        inventory.put(s, rs.getInt(2));
                    }
                }
                return inventory;
            }
        } catch (Exception e) {
            throw new WebApplicationException("Failed to get inventory: " + e.getMessage(),
                Response.Status.INTERNAL_SERVER_ERROR);
        }
    }

    private Pet mapRow(ResultSet rs) throws Exception {
        Pet pet = new Pet();
        pet.setId(rs.getObject("id", Long.class));
        pet.setName(rs.getString("name"));

        String categoryJson = rs.getString("category");
        if (categoryJson != null) {
            pet.setCategory(mapper.readValue(categoryJson, Category.class));
        }

        String photoUrlsJson = rs.getString("photo_urls");
        if (photoUrlsJson != null) {
            pet.setPhotoUrls(mapper.readValue(photoUrlsJson, new TypeReference<List<String>>() {}));
        }

        String tagsJson = rs.getString("tags");
        if (tagsJson != null) {
            pet.setTags(mapper.readValue(tagsJson, new TypeReference<List<Tag>>() {}));
        }

        String statusStr = rs.getString("status");
        if (statusStr != null) {
            pet.setStatus(Pet.StatusEnum.fromValue(statusStr));
        }

        return pet;
    }
}

File: petshop-stacks/nodejs/fastify/db/petRepository.js
'use strict';
const { query } = require('./pool');

const mapRow = (row) => ({
  id: row.id !== null ? Number(row.id) : undefined,
  name: row.name,
  category: row.category
    ? (typeof row.category === 'string' ? JSON.parse(row.category) : row.category)
    : undefined,
  photoUrls: row.photo_urls
    ? (typeof row.photo_urls === 'string' ? JSON.parse(row.photo_urls) : row.photo_urls)
    : [],
  tags: row.tags
    ? (typeof row.tags === 'string' ? JSON.parse(row.tags) : row.tags)
    : undefined,
  status: row.status || undefined,
});

const add = async (pet) => {
  let id;
  if (pet.id != null) {
    id = pet.id;
  } else {
    const { rows } = await query("SELECT nextval('pet_id_seq') AS id");
    id = Number(rows[0].id);
  }
  await query(
    `INSERT INTO pet ("id", "name", category, photo_urls, tags, status)
     VALUES ($1, $2, $3, cast($4 as json), cast($5 as json), cast($6 as pet_status))
     ON CONFLICT ("id") DO UPDATE SET "name"=EXCLUDED."name", category=EXCLUDED.category,
     photo_urls=EXCLUDED.photo_urls, tags=EXCLUDED.tags, status=EXCLUDED.status`,
    [
      id,
      pet.name,
      pet.category != null ? JSON.stringify(pet.category) : null,
      JSON.stringify(pet.photoUrls),
      pet.tags != null ? JSON.stringify(pet.tags) : null,
      pet.status || null,
    ],
  );
  return { ...pet, id };
};

const update = async (pet) => {
  if (pet.id == null) {
    const err = new Error('Pet ID is required for update');
    err.status = 400;
    throw err;
  }
  const result = await query(
    `UPDATE pet SET "name" = $1, category = $2,
     photo_urls = cast($3 as json), tags = cast($4 as json), status = cast($5 as pet_status)
     WHERE "id" = $6`,
    [
      pet.name,
      pet.category != null ? JSON.stringify(pet.category) : null,
      JSON.stringify(pet.photoUrls),
      pet.tags != null ? JSON.stringify(pet.tags) : null,
      pet.status || null,
      pet.id,
    ],
  );
  if (result.rowCount === 0) {
    const err = new Error('Pet not found');
    err.status = 404;
    throw err;
  }
  return pet;
};

const findById = async (petId) => {
  const result = await query(
    `SELECT "id", "name", category, photo_urls, tags, status::text
     FROM pet WHERE "id" = $1`,
    [petId],
  );
  if (result.rows.length === 0) {
    const err = new Error('Pet not found');
    err.status = 404;
    throw err;
  }
  return mapRow(result.rows[0]);
};

const deletePet = async (petId) => {
  await query('DELETE FROM pet WHERE "id" = $1', [petId]);
};

const findByStatus = async (status) => {
  const result = await query(
    `SELECT "id", "name", category, photo_urls, tags, status::text
     FROM pet WHERE status = cast($1 as pet_status)`,
    [status],
  );
  return result.rows.map(mapRow);
};

const findByTags = async (tags) => {
  if (!tags || tags.length === 0) return [];
  const conditions = tags.map((_, i) => `tags::jsonb @> cast($${i + 1} as jsonb)`).join(' OR ');
  const params = tags.map((t) => JSON.stringify([{ name: t }]));
  const result = await query(
    `SELECT "id", "name", category, photo_urls, tags, status::text
     FROM pet WHERE ${conditions}`,
    params,
  );
  return result.rows.map(mapRow);
};

const updateWithForm = async (petId, name, status) => {
  if (name == null && status == null) return;
  const sets = [];
  const params = [];
  if (name != null) {
    sets.push(`"name" = $${params.length + 1}`);
    params.push(name);
  }
  if (status != null) {
    sets.push(`status = cast($${params.length + 1} as pet_status)`);
    params.push(status);
  }
  params.push(petId);
  await query(`UPDATE pet SET ${sets.join(', ')} WHERE "id" = $${params.length}`, params);
};

const addPhoto = async (petId, content, contentType, metadata) => {
  const exists = await query('SELECT 1 FROM pet WHERE "id" = $1', [petId]);
  if (exists.rows.length === 0) {
    const err = new Error('Pet not found');
    err.status = 404;
    throw err;
  }
  await query(
    `INSERT INTO pet_photo ("id", pet_id, content_type, metadata, content)
     VALUES (nextval('pet_photo_id_seq'), $1, $2, $3, $4)`,
    [petId, contentType, metadata, content],
  );
  return content ? content.length : 0;
};

const getInventory = async () => {
  const result = await query(
    "SELECT status::text, cast(COUNT(*) as int) as cnt FROM pet GROUP BY status",
    [],
  );
  const inventory = {};
  for (const row of result.rows) {
    if (row.status != null) inventory[row.status] = row.cnt;
  }
  return inventory;
};

module.exports = {
  add, update, findById, deletePet, findByStatus, findByTags, updateWithForm, addPhoto, getInventory,
};
